### DISCUSSION TITLE
WinUI Community Call: September 11, 2024

### DISCUSSION THREAD

**Discussion by duncanmacmichael** [microsoft/microsoft-ui-xaml#9946] [General]
### Please join us for the September 2024 edition of the WinUI Community Call. This event will be live-streamed to YouTube: https://www.youtube.com/watch?v=3DYRxRvF-Xc&list=PLI_J2v67C23ZqsolUDaHoFkF1GKvGrttB&index=2
#
# Details
Date: Sep 11, 2024
Time: 16:00-17:00 UTC (9:00-10:00am Pacific)

Anyone and everyone is welcome - no pre-registration is required.
This will be an informal interactive live stream directly with members of our engineering team.

## Format
The community call is a call among the WinUI team that is live-streamed onto YouTube. We present on new updates, share information, welcome guests, and answer your questions.

The WinUI and WindowsAppSDK teams are excited to share an overview of 1.6 stable, AOT demos, and more! We can't wait to share what's coming next and look forward to seeing you all there!

### Agenda
- Intro
- 1.6 stable overview
- Native AOT demos
- UWP migration tips
- Plans and roadmap
- Q&A
 
### **Q&A Code of Conduct**
- To help ensure Q&A remains beneficial for the majority of the call's developer audience, please consider whether your question could be framed so that it might be more broadly relevant to our larger developer audience by ensuring it is reasonably concise, broadly beneficial, well-suited for the fast-paced live Q&A format, or able to spark open ended discussion.
- Please understand some questions are not well-suited for the fast-paced live Q&A format and should be filed as standalone [Q&A](https://github.com/microsoft/microsoft-ui-xaml/discussions/new?category=q-a&title=%5BYour+Question+Here%5D) discussions where they can be most appropriately handled. Excluding presentation topic requests, this generally includes requests for dedicated follow-ups or technically complex questions that require investigative time or expertise not included in this month's guest list.
 
### Leave us your questions in the comments on this issue or live in the YouTube chat during the stream!

**Comment by whiskhub** [comment]
Will @Rinku-Stephen also be there to present us the survey results? https://github.com/microsoft/microsoft-ui-xaml/discussions/9807#discussioncomment-10419021

**Reply by duncanmacmichael** [reply]
Yes, she will be :)

**Comment by myokeeh** [comment]
Can your team work with the Windows team to enable WinUI3 apps to be easily selectable as Kiosk apps in Windows Settings? It's very easy to do that with my current UWP app. I can't move on to using WinUI3 because of this limitation.

**Comment by lolametro** [comment]
Could you please ensure this regression (reintroduced bug in 1.6) will not be released in the final 1.6?
Even in the WinUI 3 Gallery app, this is clearly visible after upgrading from 1.5 to 1.6.

WinUI3: ContentDialog is missing entrance animation: https://github.com/microsoft/microsoft-ui-xaml/issues/8476

**Reply by lolametro** [reply]
Well, the bug was now released in the stable 1.6 release. So I guess... please fix it in the next servicing release!

**Comment by Tiktack** [comment]
WInUI3 has so many memory leaks. Are there any plans to address them at all? 
As an example #7282. Currently, WinUI never releases memory for second window and the same happens for page navigation.

**Reply by johngagefaulkner** [reply]
Is that why the WinUI 3 Gallery app constantly crashes and force-closes?

**Reply by Tiktack** [reply]
To be honest I don't remember crashes for me. But, regarding memory, if you go to Windowing -> Multiple windows and spawn additional windows and close them, allocated memory will never comeback to original. Also, every navigations to any pages allocates some memory, but not releases completely. Example below, 
Right after startup: 
![image](https://github.com/user-attachments/assets/7f0c8d42-e107-4bae-b167-9d3af25d5a64)
After maybe 50 clicks Home/All Samples(I didn't click to any other page except Home and All Samples): 
![image](https://github.com/user-attachments/assets/ca3c1b77-788c-46f5-99c3-ed783dc90ae6)

So, as you can see, I am on the same "Home" screen, but almost double the memory. And it will grow indefinitely(maybe before crash :D)

**Reply by Tiktack** [reply]
After going through almost every page, it looks like this. This is ridiculously bad. You cannot create an app that will survive a few days of usage without exiting it, cos it will eat all the memory. 
![image](https://github.com/user-attachments/assets/aa9238f3-87ef-43fc-97aa-5bf7f5f63e6f)

**Reply by johngagefaulkner** [reply]
Wow, that's worse than I expected. Out of curiosity, I tried on my PC, too:

## Initial app launch _(no navigation performed)_
<img width="1919" alt="image" src="https://github.com/user-attachments/assets/8244230f-1056-4498-b014-ac65060bf71e">

## After using app for ~10 minutes, ~30 Page Navigations
<img width="1919" alt="image" src="https://github.com/user-attachments/assets/30653647-0d9b-46b5-a5be-dd19fa16e281">

**Reply by Tiktack** [reply]
This is why we don't have somehow complex apps with WinUI3 from Microsoft. Because in a current state it's impossible. Now I understand why outlook is WebView2 app, because it's usually runs for days or even weeks. And with such memory leaks it's even worse than WebView. 
The most complex inhouse app we have is Microsoft Store, but it's UWP, not WinUI3
Another somehow complex app is "Photos", It started really well, but it was UWP. Now it's Frankenstein of WinUI3 and Webview, with weird styles across the app and really bad performance. Photo viewer app has an option to run in background to make startup  faster, it's 2024!!!

**Reply by manodasanW** [reply]
For the page navigation scenario, there was a [recent change](https://github.com/microsoft/CsWinRT/pull/1514) that might help.

**Reply by Tiktack** [reply]
From descriptions looks like it's good improvement. But issues mentioned earlier have nothing to do with GC. You can insert GC.Collect() on every page navigation but it won't make any difference at all, because it's leaked memory...

**Reply by Balkoth** [reply]
It's absolutely hilarious that the perf guy on the team is in the community call showing how to diagnose, but the framework itself being plagued by such issues.

**Reply by pjmlp** [reply]
Not only that, remember that since its early days it was touted by the team that they went with COM (WinRT is just another COM framework basically), and C++ because of performance, and WinDev being very anti-.NET since the Longhorn days.

Then one of the reasons C++/CX was replaced by C++/WinRT, was how it allowed even better performance.

And here we are, still fighting performance issues of anything WinRT not being able to match classical Win32 and .NET frameworks.

**Comment by Biyakuga** [comment]
Could you share any plans for improving performance with multiple tabs, drag-and-drop functionality, or memory usage? We would appreciate any insights or a sneak peek into what's coming.

**Comment by itayking1990** [comment]
Could you guys add input validation?

**Reply by whiskhub** [reply]
Yes please - it was once added to 1.5 experimental but then removed. What happened to it? https://github.com/microsoft/microsoft-ui-xaml/issues/179

**Comment by ghost1372** [comment]
any news for splash screen?
https://github.com/microsoft/microsoft-ui-xaml/issues/4055

**Comment by ghost1372** [comment]
- When do you plan to improve and add features to Window?
- When do you plan to provide methods for attaching UI framework content to an AppWindow?

**Reply by DarranRowe** [reply]
I don't think they ever will do anything like attaching content to an AppWindow.
AppWindow is intentionally framework independent, but what UI framework content would be attached? Whatever your answer is, how would it affect a Windows API application, WinForms or WPF application that wants to use AppWindow to make the custom titlebar implementation easier?

**Reply by ghost1372** [reply]
I should answer my question!
v1.7
 
![image](https://github.com/user-attachments/assets/83e98c89-675c-46ac-8267-ca86723e9f15)

**Comment by pjmlp** [comment]
Can we finally have a honest roadmap on the status of C++ development experience, the lack of tooling in Visual Studio for C++/WinRT on par with C++/CX, back when C++/WinRT was announced in 2016?

Especially given that C++/WinRT does not have plans to evolve beyond C++17, and apparently is unofficially in maintenance, without further improvements other than bug fixes.

**Reply by pjmlp** [reply]
Additionally, here is a talk from CppCon 2017 (!) about how C++/WinRT was supposed to deliver on the missing Visual Studio experience, caused by replacing C++/CX with C++/WinRT.

[C++/WinRT and the Future of C++ on Windows](https://youtu.be/7TdpWB_vRZM?si=6or7HbvBHqQ9NgHc)

**Comment by huoyaoyuan** [comment]
Please priortize quality, tooling support and performance over new features.
WinUI should become *the* choice of UI framework on Windows. This should the goal of the team.
I have learned for work item planning for features when I graduated to work. The current status of WinUI requires dedicated bug fixing in planning, with *much more effort* than new features.
With solid framework and vigorous ecosystem, the users and community should be able to implement more features themselves. The unreplacable job for the team is to make the framework solid.

---

Tooling is still the top pain point during my experience of the whole Appx platform, from UWP to WinUI. I have really no idea whether publishing will continue to work when I upgrade the SDK.
I customize my build in various ways, from the recommendation of .NET SDK to some experiments with new components. During debug they often just works, but publishing often fails with strange build errors. I'm already an expert of MSBuild , and that just got me started to investigate this. Sometimes things works differently from Debug to Release, from single architecture to publishing for multiple architectures. The behavior of IDE experience and CLI aren't exactly the same and makes it harder to investigate issues with CLI logging.
The latest issue I faced when upgrading to 1.6 is that some `.xaml` files are missing from `obj` output and some target wants to copy it - I have totally no idea why they are copying the source `.xaml` files instead of compiled `.xbf`! With several retries during investigation, it first successes for one or two architectures, then suddenly successes for all.

Packaging is literally on the critical path of development. It should be dealed with much more effort to ensure every path is fine. You also need to correctly award the persons responsible for it - **ensuring critical things don't regress is an impactful work, and does need a lot of effort.**
I know that they may be built by another sub-team than the UI team, but you are shipping a product together. Please inform the feedback to necessary persons and teams.

**Reply by whiskhub** [reply]
I totally agree. However I think getting "back" to feature parity with WPF or even UWP(!) is also very important. So many things that no longer work properly, that have worked in the past.

**Comment by JimmyRespawn** [comment]
Any plan for cross platform support like Xbox, Hololens etc?
Have to say quite amount of users are coming from Xbox revealed by my analytic dashboard. People love using apps on console.

**Reply by johngagefaulkner** [reply]
Wait, I thought apps on Xbox went effectively unused... Is there an untapped niche market there?

I know a few of my friends have said they wished there was a high-quality app that allowed them to view their captured Screenshots and Clips that are both on the Xbox Live network and on OneDrive (additional online storage providers would be an added bonus) and they'd happily pay $9.99 for it. 👀

**Comment by AndrewKeepCoding** [comment]
I'd like to know about the features and timeline of the upcoming `TableView` control. 🙂

**Reply by ghost1372** [reply]
Forget it, they couldn't release the simple titlebar control (which was a port of the C# version, not the 0 to 100 design) in time after months in v1.6.

**Reply by huoyaoyuan** [reply]
I hope they do validate its quality and decide to allocate more resource to fix issues, rather than releasing half-completed and forgot it.

**Comment by ReneeGA2020** [comment]
I believe the big question is what general strategy Microsoft has planned for WinUI. As I observed last year, the team seemed not to have had enough resources to carry out the necessary actions to realize WinUI’s full potential, even things like fixing core/blocking issues.

Windows users don’t really like WebUI; they prefer native apps, and we prefer them too. However, even Microsoft's own teams chose to use web to develop their apps. I think Microsoft really needs to eat its own dog food.

If Microsoft really wants to create a new generation of the Windows UI framework, I think it needs to invest more resources into the team, fix long-standing bugs, and foster better cooperation with the WinUI team and other dev teams—

For example, developing WinUI versions of apps like Word, Excel, and Visual Studio, making WinUI more integrated into Visual Studio (ahem, a designer), and by ensuring the WinUI development experience is more consistent with the .NET team’s efforts. Currently, the .NET team seems to focus on MAUI, while the Windows team seems to focus on WinUI. They are supposed to be the same on the Windows platform, but their development experiences are really disconnected.

I like many things about WinUI, but Microsoft as a whole needs to put efforts into it to ensure its success. I think Windows developers don't want to see another Silverlight, or Windows Phone, or UWP...

**Reply by huoyaoyuan** [reply]
Some smallish products are adopting WinUI 3, including PowerToys, DevHome, Photos and the explorer itself. Credits for the good start.
But it's just a small starting step. The native Windows UI must be capable to power the sold productivity applications at large scale.

**Reply by huoyaoyuan** [reply]
And that's why I still trust WPF: it's somehow old, left behind from many modernization, and evolving quite slowly. But it powers Visual Studio I use every day without causing problems. It's the trustful fallback when I don't want to trouble much with UI.

**Reply by Tiktack** [reply]
I totally agree. Still don't understand why this team understaffed. Microsoft has amazing .NET and ASP.NET teams. The same needed for WinUI. Windows Apps should be written only with WinUI, not WebView2.

**Reply by ghost1372** [reply]
When all resources should be focused on WinUI, they decide to revive WPF. In the latest version, they have implemented Fluent Design style.

**Reply by pjmlp** [reply]
@huoyaoyuan If only Photos wasn't yet another example of adding Webview2 into the mix.

**Reply by MartyIX** [reply]
> @huoyaoyuan If only Photos wasn't yet another example of adding Webview2 into the mix.

Is Photos implemented in WinUI with a WebView2 inside?

**Reply by pjmlp** [reply]
Yes, for the new image editing component, added as part of the UWP to WinUI migration. 

Note the praise for Web technology adoption on the article. 

https://blogs.windows.com/windowsdeveloper/2024/06/03/microsoft-photos-migrating-from-uwp-to-windows-app-sdk/

**Reply by huoyaoyuan** [reply]
> If only Photos wasn't yet another example of adding Webview2 into the mix.

If you browse the installation directory of Photos, the majority is written in WinUI3 and .NET 8. It is the experience that WinAppSDK tries to deliver for us. The editor part is hosted in web view, which should be integrating the work of another team.

**Reply by huoyaoyuan** [reply]
Back to the UWP time, image editing as implemented in web and hosted in WebView (Edge). They just migrated the hosting to WebView2.

**Reply by Petrarca181** [reply]
So true. To many apps switched to WebUI, this makes me very sad.

**Comment by MartyIX** [comment]
Is there a performance boost even in non-AOT scenarios in WinAppSDK 1.6?

**Comment by lolametro** [comment]
When will we be able to properly use SVGs instead of font icons?
See https://github.com/microsoft/microsoft-ui-xaml/issues/1494 for all the issues `SvgImageSource`, `PathIconSource`, ... have. 
Another issue is changing the color of SVGs (e.g. on hover, pressed, dark/light) https://github.com/microsoft/microsoft-ui-xaml/issues/507.

**Comment by ChaimCL** [comment]
I would like to confirm the progress of the inkscanvas control, will it arrive in the next version, I remember asking last year. As well as when the input validation will arrive, it's really important. Improving the stability of the framework and continuing to improve responsiveness is a priority. Delays in UI operations always give the impression of poor performance and can affect the user experience bringing a negative impression of the WinUI application.

**Comment by Gaoyifei1011** [comment]
> ### Leave us your questions in the comments on this issue or live in the YouTube chat during the stream!

Some questions to be answered by the WinUI team
1. Why not provide a window size limiting property in AppWindow? Such as MinHeight, MinWidth, MaxHeight, MaxWidth?
2. Why don't you will UIElement.ProtectedCursor types from protected to modify for the public, allowing developers to call directly, but through reflection modify the property value?
3. When to remove WinUI 3 UWP(CoreWindow) mode restrictions?
4. Is there any chance to open up the code of the contents of the AppWindow class in the future? If so, when?
5. Why not InputNonClientPointerSource right-click the events in class?
6. In the future have opportunity to SystemBackdrop (MicaBackdrop and DesktopAcrylicBackdrop) provide SystemBackdropConfiguration property?
7. Will there be opportunities to support some XAML 2009 features in WinUI 3 in the future? Such as x:Arguments, x:TypeArguments and so on.
8. When will UWP migrate from.NET Native to Modern.NET? When will WinUI 2 provide CsWinRT's Projection dll?

需要 WinUI 团队回答的一些问题
1. 为什么不在 AppWindow 提供限制窗口大小的属性？比如 MinHeight，MinWidth，MaxHeight，MaxWidth?
2. 为什么不将 UIElement.ProtectedCursor 的类型从 protected 修改为 public，从而让开发者直接调用，而是通过反射修改该属性值？
3. 什么时候移除 WinUI 3 UWP(CoreWindow) 模式的限制？
4. 未来有机会开放 AppWindow 类里面内容的代码吗？如果有，什么时候？
5. 为什么不在 InputNonClientPointerSource 类中添加右键单击的事件？
6. 未来有机会为 SystemBackdrop(MicaBackdrop 和 DesktopAcrylicBackdrop)提供 SystemBackdropConfiguration 属性吗？
7. 未来有机会能否在 WinUI 3 中支持 XAML 2009 的一些特性？比如 x:Arguments，x:TypeArguments 等等。
8. UWP 什么时候从 .NET Native 迁移到 Modern .NET？WinUI 2 什么时候提供 CsWinRT 的 Projection dll？

Integration of Windows App SDK issues in this place
Windows App SDK 问题的整合在这个地方
https://github.com/microsoft/WindowsAppSDK/issues/4699

**Comment by HEIC-to-JPEG-Dev** [comment]
I love WinUI - great work. But can you please co-ordinate releases with the other dependancy tooling. For example, the WinUI VS template took nearly a year to go from WinUI release to the VS team; and currently we have the SDK 1.6 release that needs additional components as a workaround because other teams haven't released their complimentary bits yet - the .Net SDK. It makes for a confusing release that seems buggy with temporary workarounds - not a great experience.

**Reply by whiskhub** [reply]
The 1.6 release was really problematic, as you mentioned. Especially that the .NET SDK is not available yet. The communication isn’t just underwhelming to developers; it also seems underwhelming to other teams. We can be lucky they didn't also shut down the WinUI community call, as they initially wanted to...

**Comment by huoyaoyuan** [comment]
As a framework, WinUI should avoid exposing unnecessary details to user.

After migration from UWP and WinUI, dialogs like file selection and ContentDialog are requiring hwnd or XamlRoot. I understand the technical change because there's no a single AppView bound to the UI thread, but the framework shouldn't expose this to user for simple cases. Just getting the current active window for current thread is sufficient for most cases.

**Comment by nikolayvpavlov** [comment]
Microsoft must really demonstrate real dedication to Windows App SDK and Win UI.  As we speak, I am testing the new Windows App (which will replace Azure Virtual Desktop client).  Guess what.  It is a web-app, powered by WebView2, packaged as a store app.  The UI only imitates some elements of Fluent design, the transition effects and animations are not there, and certain elements like Settings look totally out of place.  

And just don't get me started about the new Outlook or Teams.  While with Teams there is a certain idea to have web islands, the core of the app could have been at least native, but for the new Outlook there is simply no excuse.

**Comment by ghost1372** [comment]
When and in which version will the TitleBar return?
Can we see it in the next v1.6.x service update?

**Reply by ghost1372** [reply]
Fixing Bugs in v1.7
and TitleBar control possibly coming in a future 1.6 servicing release

**Comment by ward-horsfall** [comment]
Hi,

Can I request somebody ask this question (if they don't cover it).

Do they have any further details on how they will be approaching the fix of the visual designer for Visual Studio.

(Too late for me given my time zone to stay up for this live)

Thanks
Ward

**Reply by ghost1372** [reply]
wait for v1.7
![Screenshot 2024-09-11 202312](https://github.com/user-attachments/assets/3aea29a4-d43a-4c51-8d8f-3fbd25f7cb1e)

**Comment by ghost1372** [comment]
What is the purpose of creating such a discussion when none of the questions are going to be answered? You wasted our time. Worst call ever.
What is the purpose of WinUI 3 development? When you revived both WPF and now UWP?

**Reply by whiskhub** [reply]
Your comment is phrased a bit harshly, but I agree. At least half of the Community Call time should be reserved for Q&A, based on the questions presubmitted here and also chat questions. 

I understand it takes time, especially if overworked developers have to do this. However, some companies have a role called “Developer Relations Engineer.” I think this is what WinUI needs: a dedicated person who has the capacity to focus on the community.

**Reply by Tiktack** [reply]
Looks like questioning here has no point at all. The same like every Github issue, community talking with each other, but never some one from MS helps/answers.

**Comment by Panda-Sharp** [comment]
Omg omg omg I had to re-read the title 10 times... This is the best news for the last ~5 years as a #windows dev and user #uwp #notdead

**Comment by rtanabe999** [comment]
Microsoft.WindowsAppRuntime.Bootstrap.dll should be included in Windows App SDK runtime

**Comment by duncanmacmichael** [comment]
Hi everyone, thanks for all the questions! We've posted the follow-up to answer as many questions as we can that we couldn't get to during the call. Here's the link to the post:

https://github.com/microsoft/microsoft-ui-xaml/discussions/9995
